### PR TITLE
Add DVR option in neutron proposal UI

### PR DESCRIPTION
**Pull request by mjura**

### CODE REVIEW

**Comment by vuntz** [comment]
We need some javascript magic to forcefully disable the setting in case where DVR is not supported (see the validation in the model for the cases where we don't support it).

Alternatively: we validate that these cases are supported now (I think it's mostly vlan, which should work now but someone needs to check that)

**Comment by mjura** [comment]
I think that we are talking about this 
https://specs.openstack.org/openstack/neutron-specs/specs/kilo/neutron-ovs-dvr-vlan.html

**Comment by mjura** [comment]
Ok, here I have found
http://docs.openstack.org/networking-guide/scenario_dvr_ovs.html
```
In the Juno (initial) release, DVR supports VXLAN and GRE project networks. In the Kilo release, support was added for DVR with VLAN project networks. All releases support flat and VLAN external networks.
```
So it means that only `vmware` plugin is not supported. Am I right @vuntz ?

**Comment by dirkmueller** [comment]
I believe linuxbridge is also not supported with DVR (yet)

**Comment by vuntz** [comment]
> I believe linuxbridge is also not supported with DVR (yet)

"DVR now supports VLANs in addition to VXLAN/GRE" (kilo release notes)

**Comment by vuntz** [comment]
> So it means that only vmware plugin is not supported. Am I right @vuntz ?

I would rather say: only ml2 with linuxbridge or ovs is supported. If we add support for some other plugin, then I have no idea if this will work.

But adding support for dvr with vlan is possibly a bigger task; you might want to just do the javascript magic for this PR...

**Comment by mjura** [comment]
I have to discuss latest changes with @tboerger

**Comment by MaximilianMeister** [comment]
can't say much about what is supported, but I tested the javascript magic and it works :+1:

File: crowbar_framework/app/assets/javascripts/barclamps/neutron/application.js [hunk 1]
```diff
@@ -236,6 +236,7 @@ function networking_plugin_check() {
     $('#ml2_type_drivers_container').show();
     $('#ml2_type_drivers_default_provider_network_container').show();
     $('#ml2_type_drivers_default_tenant_network_container').show();
+    $('#dvr_container').show();
```
Comment (vuntz) [inline crowbar_framework/app/assets/javascripts/barclamps/neutron/application.js]: Actually, I would only show it from `ml2_type_drivers_check` if gre or vxlan is in the list of drivers.

Comment (mjura) [inline crowbar_framework/app/assets/javascripts/barclamps/neutron/application.js]: Ok, it was moved to `ml2_type_drivers_check`

**Comment by mjura** [comment]
Rebased

File: crowbar_framework/config/locales/neutron/en.yml [hunk 2]
```diff
@@ -31,6 +31,7 @@ en:
         rabbitmq_instance: 'RabbitMQ Instance'
         keystone_instance: 'Keystone Instance'
         dhcp_domain: 'DHCP Domain'
+        use_dvr: 'Enable Distributed Virtual Router'
```
Comment (dirkmueller) [inline crowbar_framework/config/locales/neutron/en.yml]: I think "Use Distributed Virtual Router (setup/layout)" would be more descriptive

Comment (mjura) [inline crowbar_framework/config/locales/neutron/en.yml]: Thank you for a tip, it was adopted

**Comment by dirkmueller** [comment]
that was not a verbatim-to-copy proposal.. you should have chosen between setup or layout, not actually copy (setup/layout) :-)

**Comment by mjura** [comment]
He he, it was already in quotes ready to use :) no worries, I was checking you.

Fixed and rebased

**Comment by dirkmueller** [comment]
+1

**Comment by mjura** [comment]
I did final tests. So DVR works fine with openvswitch ML2 driver and VLAN/GRE/VXLAN, but doesn't work with Llinux Bridge and VMware NSX.

File: crowbar_framework/app/assets/javascripts/barclamps/neutron/application.js [hunk 3]
```diff
@@ -291,6 +293,8 @@ var ovsSelection = null
 
 function ml2_mechanism_drivers_check() {
   var values = $('#ml2_mechanism_drivers').val() || [];
+  $('#dvr_container').show();
```
Comment (dirkmueller) [inline crowbar_framework/app/assets/javascripts/barclamps/neutron/application.js]: this causes flicker I guess. better move it to line 335

Comment (mjura) [inline crowbar_framework/app/assets/javascripts/barclamps/neutron/application.js]: Ok, let it be. I haven't seen any side effects of this. It was moved to 335

**Comment by mjura** [comment]
I've split up changes to different commits, should cover all usage scenarios.

**Comment by dirkmueller** [comment]
:+1:

**Comment by aplanas** [comment]
+1

**Pull request merged**